processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment;
+
+import com.google.common.collect.Lists;
+import org.apache.druid.java.util.common.granularity.Granularity;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.guava.Sequences;
+import org.apache.druid.query.QueryMetrics;
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.query.filter.InDimFilter;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.data.Indexed;
+import org.apache.druid.segment.data.ListIndexed;
+import org.apache.druid.segment.filter.AndFilter;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.Objects;
+
+public class UnnestStorageAdapter implements StorageAdapter
+{
+  private final StorageAdapter baseAdapter;
+  private final String dimensionToUnnest;
+  private final String outputColumnName;
+  private final LinkedHashSet<String> allowSet;
+
+  public UnnestStorageAdapter(
+      final StorageAdapter baseAdapter,
+      final String dimension,
+      final String outputColumnName,
+      final LinkedHashSet<String> allowSet
+  )
+  {
+    this.baseAdapter = baseAdapter;
+    this.dimensionToUnnest = dimension;
+    this.outputColumnName = outputColumnName;
+    this.allowSet = allowSet;
```

Comment:
An infilter for a MVD returns the entire row in case of any match. 
![Screen Shot 2022-11-29 at 10 44 28 PM](https://user-images.githubusercontent.com/93540295/204726426-2ed71290-5892-4df4-8641-bf92ed1c2ae2.png)

This allowSet allows to filter inside an MVD just to allow to unnest the values specified in the allowList and ignore the others